lua/randomat2/events/mathracer.lua

```diff
@@ -0,0 +1,301 @@
+CreateConVar("randomat_mathracer_kill_wrong", 1, FCVAR_NONE, "Whether to kill players who type the wrong answer")
+
+local operations = {
+    add = {
+        symbol = "+",
+        left = "addend",
+        right = "addend",
+        calc = function(a, b) return a + b end
+    },
+    sub = {
+        symbol = "-",
+        left = "minuend",
+        right = "subtrahend",
+        calc = function(a, b) return a - b end
+    },
+    mul = {
+        symbol = "×",
+        left = "factor",
+        right = "factor",
+        calc = function(a, b) return a * b end
+    },
+    div = {
+        symbol = "÷",
+        left = "dividend",
+        right = "divisor",
+        calc = function(a, b) return a / b end
+    }
+}
+
+local parentheses = {
+    base = {add = false, sub = false, mul = false, div = false},
+    addend = {add = false, sub = true, mul = false, div = false},
+    minuend = {add = false, sub = true, mul = false, div = false},
+    subtrahend = {add = true, sub = true, mul = false, div = false},
+    factor = {add = true, sub = true, mul = false, div = true},
+    dividend = {add = true, sub = true, mul = false, div = true},
+    divisor = {add = true, sub = true, mul = true, div = true}
+}
+
+local iteration
+local difficulty_options
+
+function EVENT:GenerateEquationNode(type)
+    local operation = difficulty_options["operations"][math.random(#difficulty_options["operations"])]
+    return {
+        left = {
+            value = false,
+            type = operations[operation]["left"]
+        },
+        right = {
+            value = false,
+            type = operations[operation]["right"]
+        },
+        operation = operation,
+        parentheses = parentheses[type][operation]
+    }
+end
+
+function EVENT:GenerateEquationTree(iterations, tree)
+    if iterations == 0 then
+        return tree
+    else
+        if tree then
+            local node = tree
+            local side = math.random() < 0.5 and "left" or "right"
+            while node[side]["value"] do
+                node = node[side]["value"]
+                side = math.random() < 0.5 and "left" or "right"
+            end
+            node[side]["value"] = self:GenerateEquationNode(node[side]["type"])
+            return self:GenerateEquationTree(iterations - 1, tree)
+        else
+            return self:GenerateEquationTree(iterations - 1, self:GenerateEquationNode("base"))
+        end
+    end
+end
+
+function EVENT:IsValidAnswer(answer)
+    if answer == math.huge then return false end
+    if answer == -math.huge then return false end
+    if answer ~= math.floor(answer) then return false end
+    if answer < difficulty_options["answerLowerBound"] then return false end
+    if answer > difficulty_options["answerUpperBound"] then return false end
+    return true
+end
+
+function EVENT:SolveEquationTree(tree)
+    if tree["left"]["value"] then
```

Comment:
At a quick glance the logic for the left and right traversal seems identical. Does it make sense to extract that into a function taking the direction as an argument to reduce duplication?